File: leetcode/78子集/subset.py
"""
给定一组不含重复元素的整数数组nums，返回该数组所有可能的子集（幂集）。
说明：解集不能包含重复的子集。

示例:
输入: nums = [1,2,3]
输出:
[
  [3],
  [1],
  [2],
  [1,2,3],
  [1,3],
  [2,3],
  [1,2],
  []
]
"""


class Solution(object):
    @staticmethod
    def subsets(nums: [int]) -> [[int]]:
        output = [[]]

        for num in nums:
            print(output)
            output += [curr + [num] for curr in output]

        return output


class Solution2(object):
    @staticmethod
    def subsets(nums: [int]) -> [[int]]:
        def back_track(first=0, curr=[]):
            if len(curr) == k:
                output.append(curr[:])
            for i in range(first, n):
                curr.append(nums[i])
                back_track(i + 1, curr)
                curr.pop()

        output = []
        n = len(nums)
        for k in range(n + 1):
            back_track()
        return output


if __name__ == '__main__':
    a = Solution2()
    print(a.subsets([1, 2, 3]))

File: leetcode/78子集/subset.go
package main

import (
	"fmt"
)

// func backtrack(first int, pos int, curr *[][]int) {
// 	if len(*curr) == pos {
// 		*curr = append(*curr, *curr...)
// 	}
// 	for i := first; i< len(num)
// }
// func subsets(nums []int) (output *[][]int) {
// 	n := len(nums)
// 	for k := 0; k < n+1; k++ {
// 		backtrack(0, k, output)
// 	}
// 	return output
// }
// //
// func main() {
// 	fmt.Println(subsets([]int{1, 2, 3}))
// }
func subsets(nums []int) (result [][]int) {
	result = make([][]int, 0) // result 保存最终结果
	list := make([]int, 0)    // list 保存临时结果
	backtrack(nums, 0, list, &result)
	return result
}

func backtrack(nums []int, pos int, list []int, result *[][]int) {
	ans := make([]int, len(list))
	copy(ans, list)
	*result = append(*result, ans)
	for i := pos; i < len(nums); i++ {
		list = append(list, nums[i])
		backtrack(nums, i+1, list, result)
		list = list[0 : len(list)-1]
	}
}

func subset(nums []int) (result [][]int) {
	result = make([][]int, 0)
	list := make([]int, 0) // list 保存临时结果

	for i := 0; i < len(nums)+1; i++ {
		Backtrack(0, nums, list, &result)
	}
	return result
}

func Backtrack(pos int, nums []int, list []int, result *[][]int) {

	for i := pos; i < len(nums); i++ {
		list = append(list, nums[i])
		Backtrack(i+1, nums, list, result)
		list = list[0 : len(list)-1]
	}
}

func main() {
	fmt.Println(subsets([]int{1, 2, 3}))
}
